File: GPS .py
import streamlit as st

st.title("Obtener Geolocalización en Streamlit")

html_code = """
<script>
    function getLocation() {
        if (navigator.geolocation) {
            navigator.geolocation.getCurrentPosition(
                (position) => {
                    document.getElementById("coords").innerText = 
                        "Lat: " + position.coords.latitude + 
                        ", Lon: " + position.coords.longitude;
                },
                (error) => {
                    document.getElementById("coords").innerText = "Error: " + error.message;
                }
            );
        } else {
            document.getElementById("coords").innerText = "Geolocalización no es soportada en este navegador.";
        }
    }
    window.onload = getLocation;
</script>
<div id="coords">Obteniendo ubicación...</div>
"""

st.components.v1.html(html_code)

File: static/script/gps.js

document.addEventListener("click", (e) => {
  if (e.target.id === "get-gps") {
    if (!navigator.geolocation) {
      const geo = document.querySelector("#geo");
      if (geo) geo.textContent = "Geolocalización no soportada.";
      return;
    }

    navigator.geolocation.getCurrentPosition(
      async (pos) => {
        const latitude=pos.coords.latitude;
        const longitude=pos.coords.longitude;
        try {
          const resp = await fetch("/api/gps", {
            // Los datos enviados al servidor seran procesados o modificados
            method: "POST",
            // El servidor ha de saber como interpretar el body (en json).
            // Sin dicho header se podria tratar como texto plano
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ latitud: latitude, longitud: longitude }),
          });
          await resp.json();
          await window.refreshState?.();

          const geo = document.querySelector("#geo");
          if (geo) geo.textContent="Lat: " + latitude +", Lon: " + longitude;
        } catch (e) {
          console.error("Error enviando GPS", e);
        }
      },
      (err) => {
        const geo = document.querySelector("#geo");
        if (geo) geo.textContent = "Error obteniendo ubicación: " + err.message;
      }
    );
  }
});